### PR TITLE
[IMP] account,sale,repair,website,purchase: remove setting and display correction

### PR DESCRIPTION
**Pull request by malb-odoo**
Before this commit, the accounting module had a setting on the selection of taxes used by the sales, purchase, website_sale, ... modules. In addition, the website_sale module had the same setting related to the accounting one.
This PR isolates the tax selection parameter in the website_sale module, and instead of using the tax selection, the accounting module will use rounding method (round per line, round globally).
Selecting one of the two will change the display of all account_move or other view using the tax selection setting.

Display changes:
In round per line, a column tax excluded will always be displayed while a column tax included will be in optional hide.
On the other hand, in the round globally, only the tax excluded columns will be available.

task-id: 2954332

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by Feyensv — CHANGES_REQUESTED** [review]

**Comment by KangOl** [comment]
This change doesn't make any sense to me. The rounding method is orthogonal with the mode of display of the price.

**Comment by chklop** [comment]
> This change doesn't make any sense to me. The rounding method is orthogonal with the mode of display of the price.

While the two settings made different things, they were not orthogonal and could generate displats that led to a confusion, and questions from users. The purpose of this task is to acknowledge that orthogonal settings are in this case not needed and we could prevent that confusion by linking the display to the rounding method.

The main reason is that the amounts showed on the line might be wrong depending on your rounding method. You will find examples here : https://app.excalidraw.com/l/65VNwvy7c4X/3W79uZzoq36

File: addons/website_sale/models/product_product.py [hunk 1]
```diff
@@ -14,7 +14,7 @@ class Product(models.Model):
 
     website_url = fields.Char('Website URL', compute='_compute_product_website_url', help='The full URL to access the document through the website.')
 
-    base_unit_count = fields.Float('Base Unit Count', required=True, default=1, help="Display base unit price on your eCommerce pages. Set to 0 to hide it for this product.")
+    base_unit_count = fields.Float('Base Unit Count', required=False, default=1, help="Display base unit price on your eCommerce pages. Set to 0 to hide it for this product.")
```
Comment (Williambraecky) [inline addons/website_sale/models/product_product.py:17]: Why this change?

File: addons/website_sale/models/res_config_settings.py [hunk 2]
```diff
@@ -61,7 +61,30 @@ class ResConfigSettings(models.TransientModel):
...
+        ('tax_excluded', 'Tax Excluded'),
+        ('tax_included', 'Tax Included')], string="Line Subtotals Tax Display",
+        required=True, default='tax_excluded',
+        config_parameter='website_sale.show_line_subtotals_tax_selection')
+
+    @api.depends('show_line_subtotals_tax_selection')
+    def _compute_group_show_line_subtotals(self):
+        for wizard in self:
+            wizard.group_show_line_subtotals_tax_included = wizard.show_line_subtotals_tax_selection == "tax_included"
+            wizard.group_show_line_subtotals_tax_excluded = wizard.show_line_subtotals_tax_selection == "tax_excluded"
```
Comment (Williambraecky) [inline addons/website_sale/models/res_config_settings.py:87]:
```suggestion
            wizard.group_show_line_subtotals_tax_excluded = wizard.show_line_subtotals_tax_selection == "tax_excluded"

```

**Review by Williambraecky — CHANGES_REQUESTED** [review]
Roughly read through the code, the groups should not be removed in the upgrade script but rather renamed to be owned by `website_sale` instead, as such clients won't lose their settings.

File: addons/website_sale/models/product_product.py (hunk 1 above)
Comment (malb-odoo) [reply addons/website_sale/models/product_product.py:17]: It was for a test failing about null constraint on the base_unit_count , i will try to run it on runbot to see if the change is very usefull by putting True instead of False 

> Roughly read through the code, the groups should not be removed in the upgrade script but rather renamed to be owned by `website_sale` instead, as such clients won't lose their settings.
And about that, quick question on what you advice, maybe using move_to_field in account/saas~15.5.1.2 ?

Comment (malb-odoo) [reply addons/website_sale/models/product_product.py:17]: > Roughly read through the code, the groups should not be removed in the upgrade script but rather renamed to be owned by `website_sale` instead, as such clients won't lose their settings.
And about that, quick question on what you advice, maybe using move_to_field in account/saas~15.5.1.2 ?

Comment (malb-odoo) [reply addons/website_sale/models/product_product.py]: I revert this change, wasn't useful, thx !

Comment (Williambraecky) [reply addons/website_sale/models/product_product.py]: > > Roughly read through the code, the groups should not be removed in the upgrade script but rather renamed to be owned by `website_sale` instead, as such clients won't lose their settings.
> 
> And about that, quick question on what you advice, maybe using move_to_field in account/saas~15.5.1.2 ?

You should use `rename_xmlid` in this case to set its new module. Upgrade will confirm if it's the right way to do :)

File: addons/sale/tests/common.py [hunk 3]
```diff
@@ -217,10 +216,9 @@ def setup_sale_configuration_for_company(cls, company):
                 'supplier_taxes_id': [(6, 0, [])],
             }),
         })
-
+        # rslt = super().setup_sale_configuration_for_company()
```
Comment (william-andre) [inline addons/sale/tests/common.py:219]: Uh?

File: addons/sale/tests/common.py [hunk 4]
```diff
@@ -217,10 +216,9 @@ def setup_sale_configuration_for_company(cls, company):
...
         return company_data
 
-
```
Comment (william-andre) [inline addons/sale/tests/common.py:223]: https://peps.python.org/pep-0008/#blank-lines

> Surround top-level function and class definitions with two blank lines.

File: addons/sale/tests/common.py [hunk 5]
```diff
@@ -42,7 +42,6 @@ def setUpClass(cls):
     def _enable_pricelists(cls):
         cls.env.user.groups_id += cls.env.ref('product.group_product_pricelist')
 
-
```
Comment (william-andre) [inline addons/sale/tests/common.py:45]: https://peps.python.org/pep-0008/#blank-lines

> Surround top-level function and class definitions with two blank lines.

File: addons/sale/tests/test_sale_flow.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
 # -*- coding: utf-8 -*-
+
```
Comment (william-andre) [inline addons/sale/tests/test_sale_flow.py:2]: useless diff

File: addons/account/models/account_move_line.py [hunk 7]
```diff
@@ -343,7 +343,9 @@ class AccountMoveLine(models.Model):
         digits='Discount',
         default=0.0,
     )
-
+    tax_calculation_rounding_method = fields.Selection(
+        related='company_id.tax_calculation_rounding_method',
+        string='Tax calculation rounding method', readonly=True)
```
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/models/account_move_line.py:348]: Same on all other models with lines

File: addons/repair/models/repair.py [hunk 8]
```diff
@@ -130,7 +130,9 @@ class Repair(models.Model):
     tracking = fields.Selection(string='Product Tracking', related="product_id.tracking", readonly=False)
     invoice_state = fields.Selection(string='Invoice State', related='invoice_id.state')
     priority = fields.Selection([('0', 'Normal'), ('1', 'Urgent')], default='0', string="Priority")
-
+    tax_calculation_rounding_method = fields.Selection(
+        related='company_id.tax_calculation_rounding_method',
+        string='Tax calculation rounding method', readonly=True)
```
Comment (william-andre) [inline addons/repair/models/repair.py:135]: https://peps.python.org/pep-0008/#blank-lines

> Method definitions inside a class are surrounded by a single blank line.

File: addons/repair/models/repair.py [hunk 9]
```diff
@@ -706,7 +708,9 @@ class RepairLine(models.Model):
         copy=False, readonly=True, required=True,
         help='The status of a repair line is set automatically to the one of the linked repair order.')
     tracking = fields.Selection(string='Product Tracking', related="product_id.tracking")
-
+    tax_calculation_rounding_method = fields.Selection(
+        related='company_id.tax_calculation_rounding_method',
+        string='Tax calculation rounding method', readonly=True)
```
Comment (william-andre) [inline addons/repair/models/repair.py:713]: https://peps.python.org/pep-0008/#blank-lines

> Method definitions inside a class are surrounded by a single blank line.

File: addons/website_sale/models/res_config_settings.py [hunk 10]
```diff
@@ -63,6 +63,30 @@ class ResConfigSettings(models.TransientModel):
...
+    group_show_line_subtotals_tax_included = fields.Boolean(
+        "Show line subtotals with taxes (B2C)",
+        implied_group='website_sale.group_show_line_subtotals_tax_included',
+        group='base.group_portal,base.group_user,base.group_public',
+        compute='_compute_group_show_line_subtotals', store=True, readonly=False)
+    show_line_subtotals_tax_selection = fields.Selection([
+        ('tax_excluded', 'Tax Excluded'),
+        ('tax_included', 'Tax Included')], string="Line Subtotals Tax Display",
+        required=True, default='tax_excluded',
+        config_parameter='website_sale.show_line_subtotals_tax_selection')
```
Comment (william-andre) [inline addons/website_sale/models/res_config_settings.py:83]: Where is this field displayed?

File: addons/website_sale/models/res_config_settings.py [hunk 11]
```diff
@@ -63,6 +63,30 @@ class ResConfigSettings(models.TransientModel):
...
+    group_show_line_subtotals_tax_excluded = fields.Boolean(
+        "Show line subtotals without taxes (B2B)",
+        implied_group='website_sale.group_show_line_subtotals_tax_excluded',
+        group='base.group_portal,base.group_user,base.group_public',
+        compute='_compute_group_show_line_subtotals', store=True, readonly=False)
+    group_show_line_subtotals_tax_included = fields.Boolean(
+        "Show line subtotals with taxes (B2C)",
+        implied_group='website_sale.group_show_line_subtotals_tax_included',
+        group='base.group_portal,base.group_user,base.group_public',
+        compute='_compute_group_show_line_subtotals', store=True, readonly=False)
```
Comment (william-andre) [inline addons/website_sale/models/res_config_settings.py:78]: Why do we still have to use groups?

File: addons/website_sale/security/website_sale.xml [hunk 12]
```diff
@@ -1,66 +1,67 @@
 <?xml version="1.0" encoding="utf-8"?>
-<odoo noupdate="1">
+<odoo>
+    <data noupdate="1">
```
Comment (william-andre) [inline addons/website_sale/security/website_sale.xml:3]: This is not necessary, a lot of useless diff because of this

File: addons/website_sale/models/product_product.py [hunk 13]
```diff
@@ -99,7 +99,7 @@ def _is_add_to_cart_allowed(self):
     def _get_contextual_price_tax_selection(self):
         self.ensure_one()
         price = self._get_contextual_price()
-        line_tax_type = self.env['ir.config_parameter'].sudo().get_param('account.show_line_subtotals_tax_selection')
+        line_tax_type = self.env['ir.config_parameter'].sudo().get_param('env.company.show_line_subtotals_tax_selection')
```
Comment (william-andre) [inline addons/website_sale/models/product_product.py:102]: uh

File: addons/l10n_ar/views/report_invoice.xml [hunk 14]
```diff
@@ -126,12 +125,12 @@
             <attribute name="t-value">current_subtotal + l10n_ar_values['price_subtotal']</attribute>
         </t>
         <!-- if b2c we still wants the latam subtotal -->
-        <t t-set="current_subtotal" t-value="current_subtotal + line.price_total" position="attributes">
+        <t t-set="current_total" t-value="current_subtotal + line.price_total" position="attributes">
```
Comment (william-andre) [inline addons/l10n_ar/views/report_invoice.xml:128]: Are you sure about this? Doesn't seem intuitive

File: addons/l10n_ar/views/report_invoice.xml [hunk 15]
```diff
@@ -126,12 +125,12 @@
...
             <attribute name="t-value">current_subtotal + l10n_ar_values['price_subtotal']</attribute>
         </t>
         <!-- label amount for subtotal column on b2b and b2c -->
-        <xpath expr="//th[@name='th_subtotal']/span[@groups='account.group_show_line_subtotals_tax_included']" position="replace">
-            <span groups="account.group_show_line_subtotals_tax_included">Amount</span>
+        <xpath expr="//th[@name='th_subtotal']/span" position="replace">
+            <span>Tax excl.</span>
```
Comment (william-andre) [inline addons/l10n_ar/views/report_invoice.xml:133]: Isn't this already the value in the main template?

File: addons/l10n_cl/views/report_invoice.xml [hunk 16]
```diff
@@ -156,8 +156,8 @@
             <attribute name="t-value">current_subtotal + l10n_cl_values['price_subtotal']</attribute>
         </t>
 
-        <xpath expr="//th[@name='th_subtotal']/span[@groups='account.group_show_line_subtotals_tax_included']" position="replace">
-            <span groups="account.group_show_line_subtotals_tax_included">Amount</span>
+        <xpath expr="//th[@name='th_subtotal']/span" position="replace">
+            <span>Tax excl.</span>
```
Comment (william-andre) [inline addons/l10n_cl/views/report_invoice.xml:160]: Same as for AR

File: addons/repair/models/repair.py (hunk 9 above)
Comment (william-andre) [reply addons/repair/models/repair.py:714]: Field doesn't seem necessary because there is no kanban. Which seems to be an error for mobile... (didn't test)

Comment (malb-odoo) [reply addons/repair/models/repair.py:714]: By removing the field i got an error about field does not exist in model repair.order, so it don't seem to work ?

Comment (william-andre) [reply addons/repair/models/repair.py:714]: uh?

**Review by william-andre — APPROVED** [review]

**Review by william-andre — APPROVED** [review]
@robodoo r+

**Comment by william-andre** [comment]
@robodoo r+

**Review by william-andre — APPROVED** [review]
@robodoo r+
